lib/luasocket.c

```diff
@@ -56,15 +56,32 @@ typedef struct luasocket_addr_s {
 static int luasocket_new(lua_State *L);
 static int luasocket_accept(lua_State *L);
 
+static int luasocket_checkiface(lua_State *L, int ix)
+{
+	int ifindex;
+	const char *ifname = luaL_checklstring(L, ix, NULL);
+	struct net_device *dev = dev_get_by_name(&init_net, ifname);
+	if (!dev) {
+		lua_pushinteger(L, -ENODEV);
+		lua_error(L);
+	}
+	ifindex = dev->ifindex;
+	dev_put(dev);
+	return ifindex;
+}
+
 static void luasocket_checkaddr(lua_State *L, struct socket *socket, luasocket_addr_t *addr, int ix)
 {
 	addr->family = socket->sk->sk_family;
 	if (addr->family == AF_INET) {
 		struct sockaddr_in *addr_in = (struct sockaddr_in *)addr;
 		addr_in->sin_addr.s_addr = htonl((u32)luaL_checkinteger(L, ix));
 		addr_in->sin_port = htons((u16)luaL_checkinteger(L, ix + 1));
-	}
-	else {
+	} else if (addr->family == AF_PACKET) {
+		struct sockaddr_ll *addr_ll = (struct sockaddr_ll *)addr;
+		addr_ll->sll_protocol = htons(ETH_P_ALL);
```

Comment:
shouldn't we allow users to specify the ETH proto? or AF_PACKET can only be used with `ETH_P_ALL`?

Reply:
sure, probably. good thing you spotted it, I forgot to mention. I tried to get the `protocol` from `socket->sk`, so that would be enough to specify, like:
```
local socket = socket.new(PACKET, RAW, <PROTO>)
socket:send(frame, 'eth0')
```
and it would re-use the `protocol` passed while creating the socket, but well, I failed badly. 

by the looks of it `socket->sk` does not store the `protocol` directly. it converts to all sorts of things, and it was kind of tough to unravel the code to a meaningful end, so I guessed it wouldn't hurt to pass `ETH_P_ALL` here, but it's still just a guess.

I'll look deeper and try to provide more information on this matter.

Reply:
just in case, and to make sure we're on the same page: I'm not really fond of the idea of passing the (iface, protocol) as distinct arguments to `send`. point being, it would sound odd to anyone familiar with the "classic" socket API.

The way I see, the best approach would be to re-use the value that should be already there, but as a last resort I'd vouch for something similar to what you suggested above:
```
local addr = socket:addr('eth0', ETH_P_ALL)
socket:send(frame, addr)
```

Reply:
Sorry, I believe it's safe to omit the protocol. According to the documentation:
```
         sll_protocol
                is the standard ethernet protocol type in network byte order as defined in the <linux/if_ether.h> include file.  It defaults to the socket's protocol.

```